### PR TITLE
echo.go & echo_test.go: Refactored echo pkg and added tests.

### PR DESCRIPTION
**Pull request by llogen**
Refactored the echo pkg to make it testable and then added reworked the tests in echo_test.go.

Signed-off-by: llogen <christophlange95@googlemail.com>

### CODE REVIEW

File: cmds/core/echo/echo_test.go [hunk 1]
```diff
@@ -9,32 +9,90 @@ import (
...
-		if buf.String() != v.r {
-			t.Fatalf("Want \"%v\", got \"%v\"", v.r, buf.String())
-		}
-		buf.Reset()
+	// Creating test table for test
+	for _, tt := range []struct {
+		name             string
+		input            []string
+		exp              string
+		noNewline        bool
```
Comment (rminnich) [inline cmds/core/echo/echo_test.go:19]: you can keep the flags struct and just change the way you set up the structs.
s: "..", r: '...", f: flags{noNewline: true, interpretEscapes: true,}
if you were concerned that the initializer was unreadable?

File: cmds/core/echo/echo.go [hunk 2]
```diff
@@ -35,23 +56,20 @@ func escapeString(s string) (string, error) {
 	if err != nil {
 		return "", err
 	}
-
 	return s, nil
 }
 
-func echo(f flags, w io.Writer, s ...string) error {
+func echo(w io.Writer, noNewline, interpretEscapes bool, s []string) error {
```
Comment (rminnich) [inline cmds/core/echo/echo.go:62]: I think the variadic form (s ...string) should remain. the standard echo command takes zero arguments. It should be legal to call this function with no arguments.

I think the flags type should remain. That way, if we ever add new flags, we don't have to change the function signature. Not sure what motivates removing them.

**Review by rminnich — CHANGES_REQUESTED** [review]

File: cmds/core/echo/echo.go [hunk 3]
```diff
@@ -60,42 +78,15 @@ func echo(f flags, w io.Writer, s ...string) error {
...
-    \e     escape
-    \f     form feed
-    \n     new line
-    \r     carriage return
-    \t     horizontal tab
-    \v     vertical tab
-    \0NNN  byte with octal value NNN (1 to 3 digits)
-    \xHH   byte with hexadecimal value HH (1 to 2 digits)`)
-	}
+	util.Usage(usage)
```
Comment (rminnich) [inline cmds/core/echo/echo.go:81]: oh, cool, you added another user for util.Usage() :-)

thanks for that!

File: cmds/core/echo/echo.go (hunk 2 above)
Comment (llogen) [reply cmds/core/echo/echo.go:62]: Regarding the variadic form I'm totally with you. Regarding the flags i think the most handy way is to access them directly by dereferencing, so no flag has to be passed. What do you think?

**Review by rminnich — DISMISSED** [review]

File: cmds/core/echo/echo.go (hunk 2 above)
Comment (rminnich) [reply cmds/core/echo/echo.go]: looks good! Do you think our code coverage bot can be appeased?

Comment (llogen) [reply cmds/core/echo/echo.go]: Jep, i did a minor change that appeased the coverage bot.

**Review by walterchris — DISMISSED** [review]

**Comment by ChriMarMe** [comment]
I knew the ls test of mTime would bite me in the behind.

File: cmds/core/echo/echo_test.go [hunk 4]
```diff
@@ -9,32 +9,100 @@ import (
...
+		},
+	} {
+		// Run tests
+		t.Run(tt.name, func(t *testing.T) {
+			*noNewline = tt.noNewline
+			*interpretEscapes = tt.interpretEscapes
+			*interpretBackslashEscapes = tt.interpretBackslashEscapes
+			var buf bytes.Buffer
+			if err := echo(&buf, tt.input); err != nil {
+				t.Errorf("%s", err)
```
Comment (walterchris) [inline cmds/core/echo/echo_test.go:101]: Change to match convention: echo() = %v, want %v.

**Review by walterchris — CHANGES_REQUESTED** [review]

**Review by rminnich — DISMISSED** [review]

**Review by walterchris — APPROVED** [review]

**Review by walterchris — APPROVED** [review]

**Pull request merged**